### PR TITLE
Solution for GRPC codegen deployment

### PR DESCRIPTION
**Pull request by zhangkun83**
@ejona86 please take a look
#71

### CODE REVIEW

File: compiler/README.md [hunk 1]
```diff
@@ -1,15 +1,21 @@
...
+pre-compiled binaries are not compatible with your system, you may want to
+build your own codegen.
 
-* Linux
-* The Github head of [Protobuf](https://github.com/google/protobuf) installed
-* [Gradle](https://www.gradle.org/downloads) installed
+## System requirement
 
-## Compiling and Testing the Plugin
+* Linux, or MacOSX with Clang, or Windows with MSYS2
```
Comment (ejona86) [inline compiler/README.md]: No "or" before MacOSX

Comment (ejona86) [inline compiler/README.md]: s/MacOSX/Mac OS X/

File: compiler/README.md [hunk 2]
```diff
@@ -1,15 +1,21 @@
...
+* Java 7 or up
+* [Protobuf](https://github.com/google/protobuf) 3.0.0 or up
```
Comment (ejona86) [inline compiler/README.md]: 3.0.0 isn't released. We want to say 3.0.0-alpha-2 ?

File: compiler/README.md [hunk 3]
```diff
@@ -36,3 +42,136 @@ To generate Java interfaces with protobuf nano:
...
+
+You need to have both the 32-bit and 64-bit versions of Protobuf installed.
+It's recommended that you install Protobuf to custom locations (e.g.,
+``~/protobuf-3.0.0-32`` and ``~/protobuf-3.0.0-64``) rather than the system
+default location, so that different version of Protobuf won't mess up. It
+should also be compiled statically linked.
+
+Generally speaking, to compile and install both 32-bit and 64-bit artifacts locally:
+```
+$ TARGET_ARCHS="<list-of-archs>" ../gradlew install --info
```
Comment (ejona86) [inline compiler/README.md]: Why the --info?

File: compiler/README.md [hunk 4]
```diff
@@ -36,3 +42,136 @@ To generate Java interfaces with protobuf nano:
...
+the protobuf headers and libraries. Follow the platform-specific instructions
+below.
+
+### Linux
+You must have multilib GCC installed on your system.
+
+Compile and install 32-bit protobuf:
+```
+protobuf$ CXXFLAGS="-m32" ./configure --disable-shared --prefix=$HOME/protobuf-32
+protobuf$ make clean && make -j && make install
```
Comment (ejona86) [inline compiler/README.md]: We probably don't want to have "-j" there. That will exhaust the memory of some machines if I'm reading make's documentation correctly.

File: compiler/build.gradle [hunk 5]
```diff
@@ -8,15 +9,117 @@ description = 'The protoc plugin for gRPC Java'
...
+        }
+        linker.withArguments { args ->
+          args << "-m64"
+          // Usually we need to pass "-L" args to set library search path.
+          // Clang ld has an issue when multiple versions (e.g., 32-bit and
+          // 64-bit) of the same library exist in the search paths. It will
+          // only look at the first one even if the correct one comes after.
+          // So we have to specify paths for different architectures
+          // separately through separate flags.
+          addEnvArgs("LDFLAGS_x86_64", args)
```
Comment (ejona86) [inline compiler/build.gradle]: We only care about addEnvArgs for OS X. Do we even care about 32bit on that platform? Snow Leopard (10.6) was the last version of OS X to support 32 bit, and it is unsupported as of February 2014.

File: compiler/build.gradle [hunk 6]
```diff
@@ -8,15 +9,117 @@ description = 'The protoc plugin for gRPC Java'
...
+        classpath libraries.protobuf_plugin,
+                  'com.google.gradle:osdetector-gradle-plugin:1.2.1'
     }
 }
 
-executables {
-  java_plugin {
-    baseName "$protocPluginBaseName"
+// We know how to build deployment artifacts from these host architectures
+def deployableHostArchs = ['x86_32', 'x86_64'] as HashSet
```
Comment (ejona86) [inline compiler/build.gradle]: This seems unused.

File: compiler/check-artifact.sh [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+fi
+
+# Under Cygwin, bash doesn't have these in PATH when called from Gradle which
+# runs in Windows version of Java.
+export PATH="/bin:/usr/bin:$PATH"
+
+E_PARAM_ERR=98
+E_ASSERT_FAILED=99
+
+# Usage:
```
Comment (ejona86) [inline compiler/check-artifact.sh]: Missing usage

File: compiler/check-artifact.sh [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+# Check that the codegen artifacts are of correct architecture and don't have
+# unexpected dependencies.
+# To be run from Gradle.
+# Usage: check-artifact <OS> <ARCH> <FILE>
+# <OS> and <ARCH> are ${os.detected.name} and ${os.detected.arch} from
+# osdetector-gradle-plugin
+OS=$1
+ARCH=$2
+FILE=$3
```
Comment (ejona86) [inline compiler/check-artifact.sh]: Quotes around $3 ?

File: compiler/check-artifact.sh [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+  if [[ $ret == 0 ]]; then
+    fail "found unexpected dependencies (listed above)."
+  elif [[ $ret != 1 ]]; then
+    fail "Error when checking dependencies."
+  fi  # grep returns 1 when "not found", which is what we expect
+  echo "Dependencies look good."
+  echo
+}
+
+checkArch $FILE && checkDependencies $FILE
```
Comment (ejona86) [inline compiler/check-artifact.sh]: Quotes around $FILE ?

File: compiler/build.gradle (hunk 5 above)
Comment (ejona86) [inline compiler/build.gradle]: Actually, we only care about LDFLAGS_{arch} for OS X, right?

**Comment by ejona86** [comment]
I see some of my comments are now against an outdated version.

Nice work!

**Comment by zhangkun83** [comment]
All comments have been addressed. PTAL.

**Comment by ejona86** [comment]
@zhangkun83, LGTM!

**Pull request merged**